### PR TITLE
feat(password): initial support for change password with session

### LINKED ISSUE #2313: Check TOTP during password change
The `/password/change/start` route accepts an email and password, and returns a verified keyFetchToken.  For users with TOTP enabled, it feels like we should require a TOTP code here as well - otherwise a compromise of account credentials will allow an attacker to access key material through this endpoint without doing a TOTP verification.

### PR DESCRIPTION
**Pull request by vbudhram**
Fixes #2313

### CODE REVIEW

File: lib/routes/password.js [hunk 1]
```diff
@@ -36,6 +36,206 @@ module.exports = function (
...
+      handler: function (request, reply) {
+        log.begin('Password.change', request)
+        let devicesToNotify, verifyHash, account, newKeyFetchToken, newSessionToken
+
+        const sessionToken = request.auth.credentials
+        const originatingDeviceId = sessionToken.deviceId
+        const { wrapKb, authPW}  = request.payload
+        const ip = request.app.clientAddress
+        const wantsKeys = requestHelper.wantsKeys(request)
+        const { uid } = sessionToken
```
Comment (rfk) [inline lib/routes/password.js]: Somewhere in here, we need to check that last-auth time on the session and bail out if it was too long ago, potentially with a new errno for "stale authentication" or some-such.

**Comment by vbudhram** [comment]
After speaking with Ryan, putting this back into WIP to add some additional support for some edge cases.

File: config/index.js [hunk 2]
```diff
@@ -815,6 +815,14 @@ var conf = convict({
         env: 'RECOVERY_CODE_COUNT'
       }
     }
+  },
+  changePassword: {
+    minLastAuth: {
+      doc: 'Minimum amount of time that the session has been last verified (milliseconds) 5 minute default',
```
Comment (rfk) [inline config/index.js]: Naming nits: I'm pretty sure this is a maximum not a minimum, and it's about authentication not verification.

File: docs/api.md [hunk 3]
```diff
@@ -372,7 +375,7 @@ those common validations are defined here.
 #### lib/devices
 
 * `schema`: {
-    * `id`: isA.string.length(32).regex(HEX_STRING)
+    * `id: isA.string.length(32).regex(HEX_STRING)
```
Comment (rfk) [inline docs/api.md]: It's not clear what happened to the "`" here..?

File: lib/routes/password.js [hunk 4]
```diff
@@ -29,13 +29,239 @@ module.exports = function (
...
+          }
+        }
+
+        function checkTotpToken() {
+          return totpUtils.hasTotpToken({uid})
+            .then((hasToken) => {
+              // User's that have TOTP enabled, must be in a TOTP verified session to change their password
+              if (hasToken && (sessionToken.tokenVerificationId || sessionToken.authenticatorAssuranceLevel <= 1)) {
+                throw error.unverifiedSession()
+              }
```
Comment (rfk) [inline lib/routes/password.js]: I wonder whether the TOTP verification should *also* have been done within a certain amount of time.  @vbudhram thoughts?

File: lib/routes/password.js [hunk 5]
```diff
@@ -29,13 +29,239 @@ module.exports = function (
...
+                browser: uaBrowser,
+                browserVersion: uaBrowserVersion,
+                os: uaOS,
+                osVersion: uaOSVersion,
+                deviceType: uaDeviceType,
+                formFactor: uaFormFactor
+              } = request.app.ua
+
+              // Create a sessionToken with the verification status of the current session
+              const sessionTokenOptions = {
```
Comment (rfk) [inline lib/routes/password.js]: If the originating sessionToken was TOTP verified, should this new sessionToken inherit that state?  I don't honestly know, just thinking out loud.

File: lib/routes/password.js [hunk 6]
```diff
@@ -29,13 +29,239 @@ module.exports = function (
...
   function failVerifyAttempt(passwordForgotToken) {
     return (passwordForgotToken.failAttempt()) ?
       db.deletePasswordForgotToken(passwordForgotToken) :
       db.updatePasswordForgotToken(passwordForgotToken)
   }
 
   var routes = [
+    {
+      method: 'POST',
+      path: '/password/change',
```
Comment (rfk) [inline lib/routes/password.js]: As a general nit, it feels like there is a lot of duplicated code between this route and `/password/change/finish`.  Is there a way to refactor things so they can share the bulk of the code here?

**Comment by vbudhram** [comment]
Closing this in favor of https://github.com/mozilla/fxa-auth-server/pull/2437

**Pull request closed without merging**